sqlglot/optimizer/eliminate_join_marks.py

```diff
@@ -0,0 +1,184 @@
+        Optional[exp.Join]: The join expression if the equality contains a join mark (otherwise None)
+    """
+    if not (isinstance(eq.left, exp.Column) or isinstance(eq.right, exp.Column)):
+        return None
+    new_eq = copy.deepcopy(eq)
+    left_has_join_mark = _has_join_mark(eq.left)
+    right_has_join_mark = _has_join_mark(eq.right)
+
+    if left_has_join_mark:
+        new_eq.left.set("join_mark", False)
+        assert isinstance(new_eq.left, exp.Column)
+        join_on = new_eq.left.table
+    elif right_has_join_mark:
+        new_eq.right.set("join_mark", False)
+        assert isinstance(new_eq.right, exp.Column)
+        join_on = new_eq.right.table
+    else:
+        return None
+
+    join_this = old_joins.get(join_on, old_from).this
+    return exp.Join(this=join_this, on=new_eq, kind="LEFT")
+
+
+def _eliminate_join_marks_from_select(select: exp.Select) -> exp.Select:
+    """Remove join marks from the where columns in this select statement
+    Converts them to joins and replaces any existing joins
+
+    Args:
+        node (exp.Select): The AST to remove join marks from
+
+    Returns:
+        exp.Select: The AST with join marks removed
+    """
+    scope = build_scope(select)
+    if not scope:
+        return select
+    select_scope = scope.expression
+    if not select_scope.args.get("where"):
+        return select
+    if not select_scope.args.get("joins"):
+        return select
+    old_joins: t.Dict[str, exp.Join] = {
+        join.alias_or_name: join for join in list(select_scope.args.get("joins", []))
+    }
+    new_joins: t.Dict[str, exp.Join] = {}
+    for node in find_all_in_scope(scope.expression, exp.Column):
+        #    for node in select.find_all(exp.Column):
+        if _has_join_mark(node):
+            predicate = node.find_ancestor(exp.Predicate)
+            if not isinstance(predicate, exp.Binary):
+                continue
+            predicate_parent = predicate.parent
+            join_on = predicate.pop()
+            new_join = _equality_to_join(
+                join_on, old_joins=old_joins, old_from=select_scope.args["from"]
+            )
+            if new_join:
+                new_joins = _update_join_dict(new_join, new_joins)
+            if predicate_parent:
+                _clean_binary_node(predicate_parent)
+    _update_from(select_scope, new_joins, old_joins)
+    replacement_joins = [new_joins.get(join.alias_or_name, join) for join in old_joins.values()]
+    select_scope.set("joins", replacement_joins)
+    where = select_scope.args["where"]
+    if where:
+        if not where.this:
+            select_scope.set("where", None)
+    return select
+
+
+def eliminate_join_marks(ast: exp.Expression) -> exp.Expression:
+    """Remove join marks from an expression
+
+    Args:
+        ast (exp.Expression): The AST to remove join marks from
+
+    Returns:
+        exp.Expression: The AST with join marks removed"""
+    select_nodes = list(ast.find_all(exp.Select))
```

Comment:
You should build scope here and than traverse the scope nodes.

Reply:
moved build scope to outer function, removed from inner.